<u>The <span style="color:green">**GrSyLi**</span> Package for GAP: <span style="color:green">**Gr**</span>aphs with <span style="color:green">**Sy**</span>mmetry <span style="color:green">**Li**</span>brary</u>
-----------------------------------------------------------------

The <span style="color:green">**GrSyLi**</span> package contains various collections of connected graphs with interesting symmetry properties. This includes:

  - all cubic vertex-transitive graphs on up to 1280 vertices
  - all arc-transitive 2-valent digraphs on up to 1000 vertices
  - all ($G$-)half-arc-transitive 4-valent graphs on up to 1000 vertices
  - all arc-transitive 4-valent graphs on up to 640 vertices
  - many edge-transitive 4-valent graphs on up to 512 vertices


--------------------------

The <span style="color:green">**GrSyLi**</span> package is based on the Digraphs package for GAP. The Digraph package provides lots of functionality for the inspection, construction and modification of (di)graphs, as well as various I/O capabilities.

In [1]:
LoadPackage("graphsym");;

In [2]:
NrCubicVTGraphs(50);

9

In [3]:
gamma:=CubicVTGraph(50,8);

<immutable symmetric digraph with 50 vertices, 150 edges>

In [4]:
Size(AutomorphismGroup(gamma));

300

In [5]:
Size(Stabilizer(AutomorphismGroup(gamma),1));

6

In [6]:
chi:=CharacteristicPolynomial(gamma);

x_1^50-75*x_1^48+2625*x_1^46-57025*x_1^44+862725*x_1^42-9668295*x_1^40+8334970\0*x_1^38-566676000*x_1^36+3090477525*x_1^34-13679661475*x_1^32+49537102935*x_1\^30-147486591075*x_1^28+361935120000*x_1^26-732318068400*x_1^24+1219518005475*\x_1^22-1665002761415*x_1^20+1852410909075*x_1^18-1665126487125*x_1^16+11956281\24400*x_1^14-675618038100*x_1^12+294591242055*x_1^10-96500806325*x_1^8+2284644\8625*x_1^6-3675257025*x_1^4+358338475*x_1^2-15944049

In [7]:
Collected(Factors(chi));

[ [ x_1-3, 1 ], [ x_1+3, 1 ], [ x_1^2-x_1-1, 6 ], [ x_1^2+x_1-1, 6 ],   [ x_1^4-8*x_1^2+11, 6 ] ]

In [8]:
DigraphDegeneracy(gamma);

3

In [9]:
IsPlanarDigraph(gamma);

false

In [10]:
IsHamiltonianDigraph(gamma);

true

In [11]:
IsDigraphCore(gamma);

false

In [12]:
CliqueNumber(gamma);

2

In [13]:
Splash(DotSymmetricDigraph(gamma));

If the user wants to use graphs in <span style="color:purple">**Grape**</span> format, you can simply use the command`Graph`.

In [14]:
NrCubicVTGraphs(10);

3

In [15]:
gammas:=List(AllCubicVTGraphs(10),Graph);;

In [16]:
List(gammas,ChromaticNumber);

[ 3, 2, 3 ]

In [17]:
List(gammas,IsDistanceRegular);

[ false, false, true ]

In [18]:
# Build the petersen graph as the Odd graph O_5
petersen:=Graph(SymmetricGroup(5),[[1,2]],OnSets,
                                    function(x,y) 
                                      return Intersection(x,y)=[]; 
                                    end);;

In [19]:
List(gammas,gamma->IsIsomorphicGraph(gamma,petersen));

[ false, false, true ]

<u> Stored parameters in the <span style="color:green">**GrSyLi**</span> package</u>
------------------------------------------------

The <span style="color:green">**GrSyLi**</span> package has many precomputed attributes stored for each cubic vertex-transitive graph it contains.
This includes:
 - Girth
 - Diameter
 - Cayley-ness
 - Arc-transitivity
 - Being isomorphic to a Split Praeger-Xu graph

To access this information, you can load the graph with the optional argument `data` set to `true`, or use the `SetCubicVTGraphProps` function (slow).

In [20]:
SmallCubicVTGraphsInfo(540);

  CVT: There are 259 cubic vertex-transitive graphs on 540 vertices.

  Of these 259 graphs, there are;
     - 236 bipartite graphs, 
     - 248 Cayley graphs, 
     - 0 arc-transitive graphs, 
     - 1 split Praeger-Xu graphs, 
     - 249 graphs with solvable automorphism group. 



In [21]:
gamma:=CubicVTGraph(540,10,true);

<immutable bipartite vertex-transitive symmetric digraph with bicomponent size\s 270 and 270>

In [22]:
IsBipartiteDigraph(gamma);

true

In [23]:
IsCayleyGraph(gamma);

true

In [24]:
IsSplitPraegerXuGraph(gamma);

false

All functionality is documented in the manual, which can be found in the github repository.

Once the user builds the documentation locally (see manual), they can use the GAP help viewer to find documentation on specific functions or sections.

In [25]:
?IsSplitPraegerXuGraph


graphsym: IsSplitPraegerXuGraph

<u>Example: The degree-girth problem</u>
----------------------------------------

Now let's investigate the degree-girth problem for cubic graphs.

 - The **degree-girth** problem: what is the *minimum* order of a $k$-regular graph with girth $g$?

The Moore bound gives an upper bound for the order of a graph in the degree-diameter problem, and there is also a lower bound for the degree-girth problem which we call the **girth Moore bound** $M(k,g)$.  

In [26]:
# Moore bound by girth for cubic graphs
GirthMooreBound:=function(gamma)
  local girth;

  girth:=DigraphUndirectedGirth(gamma);
  if IsEvenInt(girth) then
    return 2*(2^(girth/2)-1);
  else
    return 1+3*(2^((girth-1)/2) - 1);
  fi;
end;;

In [30]:
exact_graphs:=[];; max:=500;;

for n in [4,6..max] do
  iter:=CubicVTGraphIterator(n,true);

  for gamma in iter do

    girth:=DigraphUndirectedGirth(gamma);
    gmoore:=GirthMooreBound(gamma);

    if DigraphNrVertices(gamma)=gmoore then
      Add(exact_graphs,gamma);
    fi;
  od;
od;

Length(exact_graphs);

5

The 5 graphs meeting the girth bound are well known examples of distance-regular graphs:

  - the **tetrahedron graph** $K_4$
  - the **complete multipartite graph** $K_{3,3}$
  - the **Petersen graph**
  - the **Heawood graph**
  - the **Tutte 8-cage**

The <span style="color:blue">**AGT**</span> package (<span style="color:blue">A</span>lgebraic <span style="color:blue">G</span>raph <span style="color:blue">T</span>heory) contains specific constructors for all but the Tutte 8-cage. It can be constructed as the point-line incidence graph of the smallest generalised quadrangle $W(2)$.

In [32]:
LoadPackage("agt");;

IsIsomorphicGraph(TetrahedronGraph(),Graph(exact_graphs[1]));

true

In [33]:
IsIsomorphicGraph(CompleteMultipartiteGraph([3,3]),Graph(exact_graphs[2]));

true

In [34]:
IsIsomorphicGraph(petersen,Graph(exact_graphs[3]));

true

In [35]:
IsIsomorphicGraph(HeawoodGraph(),Graph(exact_graphs[4]));

true

In [36]:
IsIsomorphicGraph(AGT_IncidenceGraph(GeneralizedQuadrangleW(2)),Graph(exact_graphs[5]));

true

A related parameter is the **excess** of a given $k$-regular graph $\Gamma$ with girth $g$, defined as 

<center>$e(\Gamma)=|V(\Gamma)|-M(k,g)$.</center>

In particular, we are interested in how large or small the excess of a graph can be, for given $k$ and $g$.

In [39]:
excess_min:=List([1..max],x->infinity);;
excess_graphs:=List([1..max],x->[]);;

for n in [4,6..max] do
  iter:=CubicVTGraphIterator(n,true);
  
  for gamma in iter do

    girth:=DigraphUndirectedGirth(gamma);
    gbound:=GirthMooreBound(gamma);
    
    e:=DigraphNrVertices(gamma)-gbound;

    excess_min[girth]:=Minimum(excess_min[girth],e);

    if e=excess_min[girth] then
      Add(excess_graphs[girth],gamma);
    fi;
  od;
od;

In [41]:
PositionsProperty(excess_min,x->x<>infinity);
Filtered(excess_min,x->x<>infinity);

[ 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14 ]

[ 0, 0, 0, 0, 4, 0, 14, 18, 98, 36, 82, 152 ]

In [42]:
Length(excess_graphs[13]);

1

In [45]:
gamma:=excess_graphs[13][1];;
DigraphNrVertices(gamma); IsCayleyGraph(gamma);

272

true

The unique cubic vertex-transitive graph of order 272 and girth 13 is best known, showing that $202\leq n(3,13)\leq 272$.

Discovered by Hoare, shown to be smallest possible Cayley graph by Royle.



<u>Example: Cayley Neumaier graphs using the small groups library</u>
-------------------------------------------------------------

In this example, we find the smallest counterexample to Neumaier's question using the small groups library. The approach goes as follows: 

- iterate through small groups of size the desired order
- search for possible connection sets of a Cayley graph, which have a specific structure that gaurantees the existence of a regular clique
- check if such a connection set gives an erg
- build the graphs, check if they are srgs and reduce by isomorphism class

It can be shown that counterexample on at most 16 vertices must:

- have 16 vertices
- be 9-regular
- have each edge being in exactly 4 triangles
- contain a 2-regular 4-clique

In [53]:
IsInverseClosed:=function(L)
  return ForAll(L,x->x^-1 in L);
end;;

v:=16;; k:=9;; lambda:=4;; m:=2;; s:=4;; idx:=v/s;;
connection_sets:=[];;

In [54]:
for G in AllSmallGroups(v) do

  # Finding subgroups of size s up to conjugacy
  reps:=List(ConjugacyClassesSubgroups(G),Representative);;
  reps:=Filtered(reps,H->Size(H)=s);;

  # Reducing by automorphism
  aut:=AutomorphismGroup(G);
  reps:=List(Orbits(aut,List(reps,Elements),OnSets),First);

  for H in reps do
    
    # Build all possible sets T                             | General                                            : (16,9,4) case

    cos:=CosetDecomposition(G,Group(H)){[2..idx]};;         # cosets of H, not including H                       : 3 cosets 
    pairs:=List(cos,x->Combinations(x,m));;                 # m-subsets of coset representatives from each coset : 6 each
    cart:=Cartesian(pairs);;                                # all possible combinations of m-subsets             : 6^3=216 elements
    cart:=Filtered(cart,x->IsInverseClosed(Union(x)));;     # keep only inverse-closed subsets
    Ts:=List(cart,Union);;

    for T in Ts do
      if ForAll(H{[2..s]},h->Length(Intersection(T,T*h))=lambda-s+2) and
         ForAll(T,g->Length(Intersection(T,T*g))=lambda-2*m+2) then
           Add(connection_sets,[G,Union(H{[2..s]},T)]);
      fi;
    od;
  od;
od;    

In [57]:
gammas:=List(connection_sets,l->CayleyGraph(l[1],l[2]));;
gammas:=GraphIsomorphismClassRepresentatives(gammas);;

Length(gammas);

3

There are exactly 2 strongly regular graphs with these parameters, so we know we have found at least one strictly Neumaier graph.

To investigate further, we can use the <span style="color:blue">**AGT**</span> package.

In [59]:
LoadPackage("agt");;
List(gammas,IsSRG);

[ true, true, false ]

In [60]:
IsIsomorphicGraph(ComplementGraph(ShrikhandeGraph()),gammas[1]);

true

In [61]:
IsIsomorphicGraph(BilinearFormsGraph(2,2,2),gammas[2]);                  # complement of L_2(4)

true

In [62]:
IsERG(gammas[3]);

true

In [63]:
CliqueNumber(gammas[3]);

4

In [64]:
IsNG(gammas[3]);

true

In [65]:
clq:=Cliques(gammas[3],4)[1];

[ 1, 2, 5, 8 ]

In [66]:
Nexus(gammas[3],clq);

2

This approach has been applied to find thousands of new graphs, and prove the existence of graphs with unseen properties, the most notable being new possible nexi for non-strongly regular examples.